internal/cli/plugin/update.go

```diff
@@ -0,0 +1,265 @@
+	// Get the manifest from the plugin directory
+	manifest, err := plugin.GetManifestFromPluginDirectory(pluginDirectoryPath)
+	if err != nil {
+		return err
+	}
+
+	err = validateManifest(manifest)
+	if err != nil {
+		return err
+	}
+
+	// make sure that there is exactly one plugin with the same name
+	pluginCount := 0
+	for _, p := range opts.plugins {
+		if manifest.Name == p.Name {
+			pluginCount++
+		}
+	}
+	if pluginCount != 1 {
+		return fmt.Errorf(`there needs to be exactly 1 plugin with the name "%s", but there are %d`, manifest.Name, pluginCount)
+	}
+
+	// Check for duplicate commands
+	existingCommandsMap := make(map[string]bool)
+	for _, cmd := range opts.existingCommands {
+		// only add command to existing commands map if it is not part of the plugin we want to update
+		if sourcePluginName, ok := cmd.Annotations["sourcePluginName"]; !ok || sourcePluginName != manifest.Name {
+			existingCommandsMap[cmd.Name()] = true
+		}
+	}
+	if plugin.HasDuplicateCommand(manifest, existingCommandsMap) {
+		return fmt.Errorf(`could not load plugin "%s" because it contains a command that already exists in the AtlasCLI or another plugin`, manifest.Name)
+	}
+
+	return nil
+}
+
+func (opts *UpdateOpts) updatePlugin(githubAssetRelease *GithubAssetManager, existingPlugin *plugin.Plugin) error {
+	// get all plugin assets info from github repository
+	assets, err := githubAssetRelease.getPluginAssetInfo()
+	if err != nil {
+		return err
+	}
+
+	// find correct assetID using system requirements
+	assetID, err := githubAssetRelease.getAssetID(assets)
+	if err != nil {
+		return err
+	}
+
+	// download plugin asset zip file and save it as ReadCloser
+	rc, err := githubAssetRelease.getPluginAssetAsReadCloser(assetID)
+	if err != nil {
+		return err
+	}
+
+	// use the ReadCloser to save the asset zip file in the default plugin directory
+	pluginZipFilePath, err := saveReadCloserToPluginAssetZipFile(rc)
+	if err != nil {
+		return err
+	}
+	defer os.Remove(pluginZipFilePath) // delete zip file after update command finishes
+
+	// try to extract content of plugin zip file and save it in default plugin directory
+	tempPluginDirectoryName := githubAssetRelease.getPluginDirectoryName() + "_temp"
+	tempPluginDirectoryPath, err := extractPluginAssetZipFile(pluginZipFilePath, tempPluginDirectoryName)
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tempPluginDirectoryPath)
+
+	// validate the extracted plugin files
+	// if plugin is invalid, delete all of its files
+	err = opts.validatePlugin(tempPluginDirectoryPath)
+	if err != nil {
+		return err
+	}
+
+	// rename old plugin directory to <plugin-directory>_old so we can rollback in case something goes wrong
+	oldPluginDirectoryPath := existingPlugin.PluginDirectoryPath + "_old"
+	err = os.Rename(existingPlugin.PluginDirectoryPath, oldPluginDirectoryPath)
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(oldPluginDirectoryPath)
```

Comment:
Not sure if this should be here.
- Happy path: we move the old plugin, so this doesn't have to run
- Unhappy path: we move the old plugin back, so this doesn't have to run

Reply:
This is my reasoning:

- Happy path: new version of the plugin successfully downloaded, validated and moved to the right place -> we can delete the old plugin files
- Unhappy path: something goes wrong when installing the new version of the plugin -> `oldPluginDirectoryPath` will be used to rollback to the old version of the plugin ([see code here](https://github.com/mongodb/mongodb-atlas-cli/pull/3162/files/e5161f3d7f91822826b61f3c8eb7c763dbfa7798#diff-5148a859ca9e843ca53da026060c474eb50574bc1be614bb63285895f042668cR143)) so this `RemoveAll` call does nothing